### PR TITLE
Add detect_fake_mode

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98323
* __->__ #98321

This replaces fake_mode_from_tensors but it preferentially looks for
fake_mode in TracingContext and also if there is an active fake mode
on the dispatch stack, before groveling in tensors to find it.

This advances PegasusForCausalLM, which was previously failing because
we generated a graph that had a parameter (non-fake) and a SymInt,
and thus previously we failed to detect the correct fake mode.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_guards.py [hunk 1]
```diff
@@ -419,3 +419,40 @@ def make_guard(self, fn, is_volatile=False) -> Guard:
...
+            return fake_mode
+
+    from torch.utils._python_dispatch import _get_current_dispatch_mode_stack
+
+    for m in reversed(_get_current_dispatch_mode_stack()):
+        if isinstance(m, FakeTensorMode):
+            return m
+
+    fake_mode = None
+    flat_inputs, _ = tree_flatten(inputs)
```
Comment (voznesenskym) [inline torch/_guards.py:451]: what does tree_flatten return with `inputs = None`?

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -1316,23 +1317,6 @@ def stack_or_hint(t):
         ), f"Unexpected fake param {name} {stack_or_hint(param)}"
 
 
-def fake_mode_from_tensors(inputs: List[Any]):
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:1319]: RIP to a real one

**Review by voznesenskym — APPROVED** [review]

File: torch/_guards.py (hunk 1 above)
Comment (ezyang) [reply torch/_guards.py:451]: empty list

Comment (ezyang) [reply torch/_guards.py:451]: ok not empty list, but it's still fine

```
>>> from torch.utils._pytree import tree_flatten
>>> tree_flatten(None)
([None], *)
```

**Pull request closed without merging**